src/pages/proposal.tsx

```diff
@@ -57,59 +65,112 @@ import './proposals.css'
+    changingVote: false,
     confirmSubscription: false,
     confirmDeletion: false,
     confirmStatusUpdate: false,
     showVotesList: false,
     showProposalSuccessModal: false,
     showUpdateSuccessModal: false,
+    showVotingModal: false,
+    showVotingError: false,
+    showSnapshotRedirect: false,
+    retryTimer: SECONDS_FOR_VOTING_RETRY,
+    selectedChoice: EMPTY_CHOICE_SELECTION,
   })
-  const patchOptionsRef = useRef(patchOptions)
+  const [errorCounter, setErrorCounter] = useState(0)
+  const updatePageStateRef = useRef(updatePageState)
   const [account, { provider }] = useAuthContext()
   const [proposal, proposalState] = useProposal(params.get('id'))
   const { isCommittee } = useIsCommittee(account)
   const { votes, votesState } = useProposalVotes(proposal?.id)
+  const { surveyTopics, isLoadingSurveyTopics } = useSurveyTopics(proposal?.id)
   const [subscriptions, subscriptionsState] = useAsyncMemo(
     () => Governance.get().getSubscriptions(proposal!.id),
     [proposal],
     { callWithTruthyDeps: true }
   )
+  const choices: string[] = proposal?.snapshot_proposal?.choices || EMPTY_CHOICES
+  const partialResults = useMemo(() => calculateResult(choices, votes || {}), [choices, votes])
 
   const { publicUpdates, pendingUpdates, nextUpdate, currentUpdate, refetchUpdates } = useProposalUpdates(proposal?.id)
 
   const subscribed = useMemo(
     () => !!account && !!subscriptions && !!subscriptions.find((sub) => sub.user === account),
     [account, subscriptions]
   )
-  const [voting, vote] = useAsyncTask(
-    async (_: string, choiceIndex: number) => {
-      if (proposal && account && provider && votes) {
+
+  const proposalResults = useRef<HTMLDivElement>(null)
+
+  const handleScrollClick = () => {
+    proposalResults.current?.scrollIntoView({ behavior: 'smooth' })
+  }
+
+  const [castingVote, castVote] = useAsyncTask(
+    async (selectedChoice: SelectedChoice, survey?: Survey) => {
+      if (proposal && account && provider && votes && selectedChoice.choiceIndex) {
         const web3Provider = new Web3Provider(provider)
         const [listedAccount] = await web3Provider.listAccounts()
-        await SnapshotApi.get().castVote(web3Provider, listedAccount, proposal.snapshot_id, choiceIndex)
-        patchOptions({ changing: false, confirmSubscription: !votes[account] })
-        votesState.reload()
+        try {
+          await SnapshotApi.get().castVote(
+            web3Provider,
+            listedAccount,
+            proposal.snapshot_id,
+            selectedChoice.choiceIndex!,
+            SurveyEncoder.encode(survey)
+          )
+          updatePageState({
+            changingVote: false,
+            showVotingModal: false,
+            showVotingError: false,
+            confirmSubscription: !votes![account!],
+          })
+          votesState.reload()
+        } catch (err) {
+          console.error(err, { ...(err as Error) }) //TODO report error
```

Comment:
If we are leaving this todo please write it like `// TODO: report error` so it is findable in a quick search 😃